lib/reporting/sp_proofing_events_by_uuid.rb

```diff
@@ -0,0 +1,205 @@
+        as_csv.each do |row|
+          csv << row
+        end
+      end
+    end
+
+    def as_emailable_reports
+      [
+        EmailableReport.new(
+          title: "#{agency_abbreviation} Proofing Events By UUID",
+          table: as_csv,
+          filename: "#{agency_abbreviation.downcase}_proofing_events_by_uuid",
+        ),
+      ]
+    end
+
+    def csv_header
+      [
+        'UUID',
+        'Workflow Started',
+        'Documnet Capture Started',
+        'Document Captured',
+        'Selfie Captured',
+        'Document Authentication Passed',
+        'SSN Submitted',
+        'Personal Information Submitted',
+        'Personal Information Verified',
+        'Phone Submitted',
+        'Phone Verified',
+        'Verification Workflow Complete',
+        'Identity Verified for In-Band Users',
+        'Identity Verified for Verify-By-Mail Users',
+        'Identity Verified for Fraud Review Users',
+        'Out-of-Band Verification Pending Seconds',
+        'Agency Handoff Visited',
+        'Agency Handoff Submitted',
+      ]
+    end
+
+    def data
+      @data ||= fetch_results.map do |result_row|
+        process_result_row(result_row)
+      end
+    end
+
+    def process_result_row(result_row)
+      login_uuid = result_row['login_uuid']
+      agency_uuid = convert_uuid(login_uuid)
+      return unless agency_uuid.present?
+      [
+        agency_uuid,
+        result_row['workflow_started'] == '1',
+        result_row['doc_auth_started'] == '1',
+        result_row['document_captured'] == '1',
+        result_row['selfie_captured'] == '1',
+        result_row['doc_auth_passed'] == '1',
+        result_row['ssn_submitted'] == '1',
+        result_row['personal_info_submitted'] == '1',
+        result_row['personal_info_verified'] == '1',
+        result_row['phone_submitted'] == '1',
+        result_row['phone_verified'] == '1',
+        result_row['online_workflow_completed'] == '1',
+        result_row['verified_in_band'] == '1',
+        result_row['verified_by_mail'] == '1',
+        result_row['verified_fraud_review'] == '1',
+        result_row['out_of_band_verification_pending_seconds'].to_i,
+        result_row['agency_handoff'] == '1',
+        result_row['sp_redirect'] == '1',
+      ]
+    end
+
+    def convert_uuid(uuid)
+      user = User.find_by(uuid: uuid)
+      user&.agency_identities&.find_by(agency:)&.uuid
+    end
+
+    def agency
+      @agency ||= begin
+        record = Agency.find_by(abbreviation: agency_abbreviation)
+        raise "Unable to find agency with abbreviation: #{agency_abbreviation}" if record.nil?
+        record
+      end
+    end
+
+    def fetch_results
+      cloudwatch_client.fetch(
+        query:,
+        from: time_range.begin.beginning_of_day,
+        to: time_range.end.end_of_day,
+      )
+    end
+
+    def cloudwatch_client
+      @cloudwatch_client ||= Reporting::CloudwatchClient.new(
+        num_threads: 1,
+        ensure_complete_logs: false,
```

Comment:
just in case there are ever more than 10k users in a time period, maybe we add a warning if the response size is exactly 10k?

Reply:
Yeah, probably not a bad idea. I also wonder if we could sort by UUID and run another query for UUIDs after the last one if the result is exactly 10k?

Reply:
Since there should only be one row per UUID, maybe easier to just slice the input UUIDS into batches of 10k?

```
uuids.each_slice(10_000) do |slice|
  # ... run queries
end
```